pkg/validate/scoped/hydrate/namespace_hydrator.go

```diff
@@ -15,85 +15,148 @@
 // NamespaceSelectors since they are no longer needed after this point.
-func buildSelectorMap(objs *objects.Scoped) (map[string][]string, status.MultiError) {
-	var namespaces, nsSelectors, others []ast.FileObject
+func buildSelectorMap(ctx context.Context, c client.Client, objs *objects.Scoped) (map[string][]string, status.MultiError) {
+	var namespaces, others []ast.FileObject
+	nsSelectorMap := make(map[string]labels.Selector)
+	selectedNamespaceMap := make(map[string][]labels.Selector)
+	var errs status.MultiError
+	hasDynamicNSSelector := false
+
 	for _, obj := range objs.Cluster {
 		switch obj.GetObjectKind().GroupVersionKind() {
 		case kinds.Namespace():
 			namespaces = append(namespaces, obj)
 		case kinds.NamespaceSelector():
-			nsSelectors = append(nsSelectors, obj)
+			nsSelector, err := convertToNamespaceSelector(obj)
+			if err != nil {
+				errs = status.Append(errs, err)
+				continue
+			}
+			selector, err := labelSelector(nsSelector)
+			if err != nil {
+				errs = status.Append(errs, err)
+				continue
+			}
+			nsSelectorMap[nsSelector.Name] = selector
+
+			if nsSelector.Spec.Mode == v1.NSSelectorDynamicMode {
+				hasDynamicNSSelector = true
+			}
 		default:
 			others = append(others, obj)
 		}
 	}
 
-	var errs status.MultiError
-	selectorMap := make(map[string][]string)
-
-	for _, obj := range nsSelectors {
-		var selected []string
-		selector, err := labelSelector(obj)
-		if err != nil {
+	if objs.AllowAPICall && objs.DynamicNSSelectorEnabled != hasDynamicNSSelector {
+		// The DynamicNSSelectorEnabled state mismatches with the required state, set the
+		// annotation so that the reconciler-manager will recreate the reconciler
+		// with the correct setting for the Namespace watch.
+		if err := setDynamicNSSelectorEnabled(ctx, c, objs.SyncName, hasDynamicNSSelector); err != nil {
 			errs = status.Append(errs, err)
-			continue
+			return nil, errs
 		}
+	}
 
+	nsList := &corev1.NamespaceList{}
+	// Fetch on-cluster Namespaces if allowing API calls and hasDynamicNSSelector,
+	// so that the NamespaceSelector can select resources matching on-cluster Namespaces.
+	if objs.AllowAPICall && hasDynamicNSSelector {
+		if err := c.List(ctx, nsList); err != nil {
+			errs = status.Append(errs, selectors.ListNamespaceError(err))
+			return nil, errs
+		}
+	}
+
+	selectorMap := make(map[string][]string)
+	for nsSelector, selector := range nsSelectorMap {
+		var selectedNamespaces []string
+		// Select dynamic/on-cluster Namespaces that match the NamespaceSelector's labels.
+		for _, ns := range nsList.Items {
+			if selector.Matches(labels.Set(ns.GetLabels())) {
+				selectedNamespaces = append(selectedNamespaces, ns.GetName())
+				// selectedNamespaceMap only stores dynamic Namespaces.
+				selectedNamespaceMap[ns.GetName()] = append(selectedNamespaceMap[ns.GetName()], selector)
+			}
+		}
+
+		// Select statically declared Namespaces that match the NamespaceSelector's labels.
```

Comment:
Does `static` refer to a namespace that is declared in the sync source?

Reply:
Correct
- `static` only selects resources matching Namespaces statically declared in the source of truth.
- `dynamic` selects both statically declared Namespaces and on-cluster Namespaces.

That is explained in the CRD.